kotlinx-coroutines-core/common/src/internal/DispatchedTask.kt

```diff
@@ -102,11 +103,14 @@ internal abstract class DispatchedTask<in T> internal constructor(
                     }
                 }
             }
+        } catch (e: DispatchException) {
+            dispatchException = e
         } catch (e: Throwable) {
             // This instead of runCatching to have nicer stacktrace and debug experience
             fatalException = e
         } finally {
             fatalException?.let { handleFatalException(it) }
+            dispatchException?.let { handleCoroutineException(delegate.context, it.cause) }
         }
```

Comment:
I just noticed: does this even need a `finally` now? I think this logic can just be moved to the `catch` blocks, can't it?